core/arch/arm/mm/core_mmu.c

```diff
@@ -908,6 +908,14 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 			map->attr = core_mmu_type_to_attr(map->type);
 			va -= map->size;
 			va = ROUNDDOWN(va, map->region_size);
+			/*
+			 * Make sure that va is aligned with pa for
+			 * efficient pgdir mapping. Basically pa &
+			 * pgdir_mask should be == va & pgdir_mask
+			 */
+			if (map->size > 2 * CORE_MMU_PGDIR_SIZE)
+				va -= CORE_MMU_PGDIR_SIZE -
+					((map->pa - va) & CORE_MMU_PGDIR_MASK);
```

Comment:
This will introduce holes in the mapping and  potentially wasted 4kB mapping table entries. Looks related to https://github.com/OP-TEE/optee_os/pull/2376/commits/4cba86b0b0df0195f1ee8aeca7f663bec8b4c62e.

Reply:
This makes more efficient use of pagetables. 2M mapping will be used where possible, instead of large quantity of 4k mapping.

Reply:
It looks like this logic is only activated when we're actually saving page tables.